pkg/controller/acmeorders/sync.go

```diff
@@ -152,6 +152,15 @@ func (c *controller) Sync(ctx context.Context, o *cmacme.Order) (err error) {
 		return err
 	}
 
+	if o.Status.State == cmacme.Ready {
```

Comment:
This could potentially be a problem for us in the case where:

1) we go to finalize an order
2) the order is finalized with the ACME server (so it moves to a 'valid' state)
3) we fail to update the `status` block of the Order to say it's valid (for whatever reason)

This will result in is continually retrying finalization with an error stating "order is in an invalid state for finalization" (or words to that effect)

I see we do attempt to handle this case here: https://github.com/jetstack/cert-manager/blob/995c7fd3c3b84e8e41678dfe4e30003c8aae9ce5/pkg/controller/acmeorders/sync.go#L519-L534

however, I'm not 100% sure whether this case would be reached due to: https://github.com/jetstack/cert-manager/blob/995c7fd3c3b84e8e41678dfe4e30003c8aae9ce5/pkg/controller/acmeorders/sync.go#L509-L518

Reply:
> This will result in is continually retrying finalization with an error stating "order is in an invalid state for finalization" (or words to that effect)

I think what would happen in that case would be:
1. We submit finalize the order first time, miss to update the state correctly, return an error from reconciler
2. Attempt to finalize second time where the second finalization would result in a 403 error from ACME (from [here-ish](https://github.com/letsencrypt/boulder/blob/main/wfe2/wfe.go#L2280-L2288)) because at that point the ACME order would already be 'valid' 
3. The 403 would result in us setting the cert-manager `Order` CR's status to 'errored' [here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/acmeorders/sync.go#L502) and return nil
4. We don't make any more ACME calls for an errored `Order` CR ([here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/acmeorders/sync.go#L87)), the `CertificateRequest` would be marked as failed and  a new `CertificateRequest` and a new `Order` would be created after 1 hour backoff

This probably isn't ideal, perhaps we could somehow check if the actual ACME Order isn't already finalized no matter of the `Order` CR's status before calling finalize, but I _think_ this would not result in continuous reconcile

Reply:
👍 I definitely think we should fix this so we don't recreate the order from scratch though - in certain cases where conflicts are more likely, users could quickly hit rate limits (plus, if it's happened once it _may_ be more likely to happen again..)

Reply:
Agree that it should be fixed, I spent a bit of time looking into this and I think there are more related issues, for example if we hit [this](https://github.com/jetstack/cert-manager/blob/725705cde5f0c155664fd5a6ee4e23627e95a854/pkg/controller/acmeorders/sync.go#L109) we don't fetch alternative certificate chains.
Since those are both pre-existing issues and aren't caused by this PR, I'd like to fix them in separate PRs to keep this one small-ish and increase the chance for it to get merged in 1.7 :)

Reply:
I've fixed this in https://github.com/jetstack/cert-manager/pull/4697